jsoncdb/org.eclipse.cdt.jsoncdb.freescale/doc/compiler.xhtml

```diff
@@ -0,0 +1,79 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
+<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
+  <head>
+    <title>Freescale Compilers</title>
+  </head>
+  <body>
+    <div id="lsp">
+      <h4>Freescale Compilers</h4>
+      <table style="border-collapse:collapse">
+        <thead>
+          <tr>
+            <th>Compiler</th>
+            <th>Vendor/Standard</th>
+            <th>Language</th>
+            <th>Macros</th>
+            <th>Include directories</th>
+            <th>System include
+              <br />
+              directories
+            </th>
+            <th>Compiler executables</th>
+          </tr>
+        </thead>
+        <tbody>
+  <tr>
+  <td>chc</td>
+  <td>Freescale / NXP</td>
+  <td>C</td>
+  <td>Yes</td>
+  <td>Yes</td>
+  <td>Yes</td>
+  <td>chc08, chc12.exe</td>
+  </tr>
+        </tbody>
+      </table>
+    </div>
+
+    <div id="builtins">
+      <h4>Freescale Compilers</h4>
+      <table style="border-collapse:collapse">
+        <thead>
+          <tr>
+            <th>Compiler</th>
+            <th>Vendor</th>
+            <th>
+              System
+              <br />
+              include paths
+            </th>
+            <th>
+              Local
+              <br />
```

Comment:
This doc should indicate that the parser supports (local) include paths, e.g. `-I /path/to/header.h`

and system-wide include pahts with `-EnvLIBPATH=${path}`: `-Env<Environment Variable>=<Variable Setting>` 
> [...] If the file is not found, the search continues with GENPATH, LIBPATH and LIBRARYPATH for header files in double quotes

How should the line read to claim above cases?

Reply:
Simply delete the tags `<div id="builtins">` and `<div id="builtins_enhanced">` from this file as well as the references in file `help_content_extension.xml` (tag `<contribution content="doc/compiler.xhtml#builtins_enhanced"...`).

Anyway, you might have misunderstood my term _compiler-built-ins detection_. This refers to the macros and include paths that are more or less hardcoded into the compiler executable. For example, gcc spits out these values when it is given the arguments `"-E", "-P", "-dM", "-Wp,-v"`.
If the freescale compiler has options to make it spit out its built-in values, pls add code to support it. 
It is very helpful for users if they are not forced to specify the values manually in the _CDT Managed Build Setting Entries_ provider because

1. Built-In include paths may be **different per machine**, which is not very team/SCM friendly.
2. Users of CDT's cmake core build project type have no chance to specify the values; core build projects do not offer a way to specify these.

I would help to implents built-in detection but could not find an manual for the compiler on the web.

Reply:
This should be the doc:
https://www.nxp.com/docs/en/reference-manual/HC12COMPILERRM.pdf

But i can't find any reference to compiler built ins

Reply:
ah wait, seems like e.g. `-LpCfg` is needed, which will output `c:\lib\hc12c\include\stdlib.h` instead of `stdlib.h`, right?

Reply:
From their docs:
_The -LpCfg option specifies how source file and -line information is formatted in the preprocessor output_

Reply:
`-Ldf` it is, i'll implement parsing of the output. Is there still something useful missing?

<details>
<summary>predef.h</summary>

```C
/* resolved by preprocessor: __LINE__ */
/* resolved by preprocessor: __FILE__ */
/* resolved by preprocessor: __DATE__ */
/* resolved by preprocessor: __TIME__ */
#define __STDC__ 0
#define __VERSION__ 5046
#define __VERSION_STR__ "V-5.0.46"
#define __MWERKS__ 1
#define __BITFIELD_LSBIT_FIRST__
#define __BITFIELD_MSBYTE_FIRST__
#define __BITFIELD_MSWORD_FIRST__
#define __BITFIELD_TYPE_SIZE_REDUCTION__
#define __SMALL__
#define __HC12__
#define __NOT_PROCESSOR_X4__
#define __PTR_SIZE_2__
#define __DIRECT_ADR__ 0
#define __CONST_PTR_SIZE_2__
#define __NON_CONST_PTR_SIZE_2__
#define __NO_DPAGE__
#define __DPAGE_ADR__ 0x0034
#define __NO_PPAGE__
#define __PPAGE_ADR__ 0x0035
#define __NO_EPAGE__
#define __EPAGE_ADR__ 0x0036
#define __NO_GPAGE__
#define __GPAGE_ADR__ 0x0010
#define __NO_RPAGE__
#define __RPAGE_ADR__ 0x0016
#define __FLOAT_IS_IEEE32__
#define __DOUBLE_IS_IEEE32__
#define __LONG_DOUBLE_IS_IEEE32__
#define __LONG_LONG_DOUBLE_IS_IEEE32__
#define __CHAR_IS_SIGNED__
#define __CHAR_IS_8BIT__
#define __SHORT_IS_16BIT__
#define __INT_IS_16BIT__
#define __ENUM_IS_16BIT__
#define __LONG_IS_32BIT__
#define __LONG_LONG_IS_32BIT__
#define __SIZE_T_IS_UINT__
#define __PTRDIFF_T_IS_INT__
#define __ENUM_IS_SIGNED__
#define __VTAB_DELTA_IS_16BIT__
#define __WCHAR_T_IS_UCHAR__
#define __PTRMBR_OFFSET_IS_16BIT__
#define __PLAIN_BITFIELD_IS_SIGNED__
#define __ELF_OBJECT_FILE_FORMAT__
#define __PRODUCT_HICROSS_PLUS__
#define __HIWARE__
#define __ARCHIMEDES__
#define __OPTIMIZE_FOR_SIZE__
#define __BIG_ENDIAN__
#define __DEMO_MODE__
```
</details>

Reply:
built-in include paths? On linux, this is most of the time `/usr/include` unless you are using a cross-compiler. On windows, you never know where the compiler and its header files are located.

Reply:
didn't saw a built-in for that, instead system includes are given by `-Env"LibPath=..."`, at least i didn't saw anything else.

But for the built-ins output itself, how can i cleanly implement parsing of such an output file? 

All i saw is based on parsing output to `stdout/-err`. Also `processLine` is only called if there is actuall output to  `stdout/-err` (`new OutputSniffer(bopOut, console == null ? null : console.getOutputStream(), result),`)

Reply:
> But for the built-ins output itself, how can i cleanly implement parsing of such an output file?
> 
> All i saw is based on parsing output to `stdout/-err`. Also `processLine` is only called if there is actuall output to `stdout/-err` (`new OutputSniffer(bopOut, console == null ? null : console.getOutputStream(), result),`)

You will read the `predef.h` file from java and have to write a parser yourself. Or maybe you can instruct the compiler to output to stderr instead of `predef.h`.

Reply:
Still i don't see how the parser is called if there is no output at all. Instead i added the possibility to read and process a file within the `CompilerBuiltinsDetector`, while not touching the API regarding the target filename and instead using a pre defined filename (`built_ins.h`). Is both ok?

Reply:
In the workbench preferences on the _JSON Compilation Database Parser_ you can enable a console.
It will at least show the command-line.

Reply:
New approach on reading the built-ins file by inhertiting from ` IBuiltinsDetectionBehavior` and adding a getter for the output file. Class `FreescaleBuiltinsDetector` is than using the getter to open and parse the output file. Did some copy/pasta with your sources, hope this is ok.

Is this an appropriate solution to parse the built-ins file?